# Практическая работа №2: Обработка выборочных данных. Нахождение точечных оценок параметров распределения
Выполнили студенты гр. 2381 Комосский Егор Артемович и Рыжиков Иван Сергеевич. Вариант №17.

## Цель работы
Получение практических навыков нахождения точечных статистических оценок параметров распределения.

## Основные теоретические положения

**Начальный эмпирический момент**:
$$
\bar M_k=\frac{1}{N}\sum n_j\,x_j^{\,k}
$$

**Центральный эмпирический момент**:
$$
\bar m_k=\frac{1}{N}\sum n_j\,(x_j-\bar x_в)^k,
где\ \bar{x}_в = \bar{M}_1.
$$

**Выборочная дисперсия**:
$$
D_в=\bar m_2=\frac{1}{N}\sum n_j\,(x_j-\bar x_в)^2
$$

**Выборочное СКО**:
$$
\sigma_в = \sqrt{D_в}
$$

**Исправленная выборочная дисперсия**:
$$
s^2=\frac{N}{N-1}\,D_в
$$

**Исправленное СКО**:
$$
s=\sqrt{s^2}
$$

**Статистическая оценка асимметрии**:
$$
\bar A_s=\frac{\bar m_3}{s^3}
$$

**Статистическая оценка эксцесса**:
$$
\bar E=\frac{\bar m_4}{s^4}-3
$$

**Условная варианта**:
$$u_j = \frac{x_j - C}{h},$$
где $C$ − условный ноль, значение которого выбирается равным значению варианты интервального ряда, являющейся средней или близкой к средней по значению в этом ряду.

**Условный момент k-го порядка**:
$$
\bar{M}_k^* = \frac{1}{N}\sum n_j\left( \frac{x_j-C}{h} \right)^k = \frac{1}{N}\sum n_j u_j^k
$$

**Мода вариационного ряда** - варианта, которой соответствует наибольшая частота:
$$
M_o^* = x_{i-1} + h \cdot \frac{\tilde{m}_i - \tilde{m}_{i-1}}{(\tilde{m}_i - \tilde{m}_{i-1})+(\tilde{m}_i - \tilde{m}_{i+1})},
$$
где $h$ − длина интервалов; $\tilde{m}_i$ − относительная частота i-го интервала.

**Медиана** - варианта, соответсвующая середине ранжированного ряда:
$$
M_e^* = x_{i-1} + \frac{h}{\tilde{m}_i} \left( 0.5 - \sum_{j=1}^{i-1} \tilde{m}_j \right)
$$

**Выборочный коэффициент вариации** - характеризует разброс значений вариант относительно выборочного среднего:
$$
V^* = \frac{\sigma_в}{\bar x_в} \cdot 100\%
$$

## Постановка задачи

Для заданных выборочных данных вычислить с использованием метода моментов и условных вариант точечные статистические оценки математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации исследуемой случайной величины. Полученные результаты содержательно проинтерпретировать.

## Порядок выполнения работы


1. Для середин интервального ряда, полученного в практической работе №1, вычислить условные варианты. Заполнить табл. 1 (в последней строке $\sum$ необходимо заполнить суммы столбцов; ячейки отмеченные прочерком заполнять не надо). Провести контроль вычислений.
2. Вычислить условные эмпирические моменты $\bar M^{*}_k$ через условные варианты. С помощью условных эмпирических моментов вычислить центральные эмпирические моменты $\bar m_k$. Полученные результаты занести в табл. 2.
3. Вычислить выборочные среднее $\bar x_в$ и дисперсию $D_в$ с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Вычислить выборочное СКО $\sigma_в$.
4. Вычислить исправленную выборочную дисперсию $s^2$ и исправленное СКО $s$. Сравнить данные оценки с смещёнными оценками дисперсии и СКО. Сделать выводы.
5. Найти статистическую оценку коэффициентов асимметрии $\bar A_s$
 и эксцесса $\bar E$. Сделать выводы.
6. Для интервального ряда вычислить моду $M^{*}_o$, медиану $M^{*}_e$ и коэффициент вариации $V^{*}$ заданного распределения. Сделать выводы.


In [ ]:

import numpy as np
import pandas as pd
import math


a = np.array([0.03, 0.053125, 0.07625, 0.099375, 0.1225, 0.145625, 0.16875, 0.191875])    # нижние границы
b = np.array([0.053125, 0.07625, 0.099375, 0.1225, 0.145625, 0.16875, 0.191875, 0.215])   # верхние границы
n_i = np.array([2, 2, 16, 12, 29, 19, 21, 7])                                             # частоты
N = int(n_i.sum())
N


108

## Шаг 1. Середины интервалов и условные варианты
Берём середины $x_i = (a_i+b_i)/2$. Ширина интервала $h$ (если ряд равномерный, то одинаковая).

Условные варианты:
$$u_i=\frac{x_i-C}{h},$$
где $C$ — условное начало отсчёта.

In [ ]:

# Середины интервалов
x_i = (a + b) / 2

# Шаг (ширина интервала)
h = float((b - a)[0])

# Условное начало отсчёта A: по умолчанию возьмём середину интервала с максимальной частотой
idx_modal = int(np.argmax(n_i))
C = float(x_i[idx_modal])

u_i = (x_i - C) / h

pd.DataFrame({
    "i": np.arange(1, len(x_i)+1),
    "a": a, "b": b,
    "x_i": x_i,
    "n_i": n_i,
    "u_i": u_i
})


,i,a,b,x_i,n_i,u_i
0,1,0.030000,0.053125,0.041563,2,-4.0
1,2,0.053125,0.076250,0.064687,2,-3.0
2,3,0.076250,0.099375,0.087813,16,-2.0
3,4,0.099375,0.122500,0.110937,12,-1.0
4,5,0.122500,0.145625,0.134063,29,0.0
5,6,0.145625,0.168750,0.157188,19,1.0
6,7,0.168750,0.191875,0.180312,21,2.0
7,8,0.191875,0.215000,0.203437,7,3.0


## Таблица №1
Считаем суммы $\sum n_i u_i^k$ (для $k=1..4$) и вспомогательные столбцы.

In [ ]:

t1 = pd.DataFrame({
    "i": np.arange(1, len(x_i)+1),
    "x_i": x_i,
    "n_i": n_i,
    "u_i": u_i,
})

t1["n_i*u_i"]   = t1["n_i"] * t1["u_i"]
t1["n_i*u_i^2"] = t1["n_i"] * (t1["u_i"]**2)
t1["n_i*u_i^3"] = t1["n_i"] * (t1["u_i"]**3)
t1["n_i*u_i^4"] = t1["n_i"] * (t1["u_i"]**4)
t1["n_i*(u_i+1)^4"] = t1["n_i"] * ((t1["u_i"] + 1)**4)

# Суммы
sum_row = t1[["n_i", "n_i*u_i", "n_i*u_i^2", "n_i*u_i^3", "n_i*u_i^4", "n_i*(u_i+1)^4"]].sum()
sum_row.name = "Σ"

t1_display = t1.copy()
t1_display.loc["Σ", "i"] = ""
t1_display.loc["Σ", "x_i"] = ""
t1_display.loc["Σ", "n_i"] = sum_row["n_i"]
t1_display.loc["Σ", "u_i"] = "-"
t1_display.loc["Σ", "n_i*u_i"] = sum_row["n_i*u_i"]
t1_display.loc["Σ", "n_i*u_i^2"] = sum_row["n_i*u_i^2"]
t1_display.loc["Σ", "n_i*u_i^3"] = sum_row["n_i*u_i^3"]
t1_display.loc["Σ", "n_i*u_i^4"] = sum_row["n_i*u_i^4"]
t1_display.loc["Σ", "n_i*(u_i+1)^4"] = sum_row["n_i*(u_i+1)^4"]

t1_display


/tmp/ipykernel_181/2444747941.py:19: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  t1_display.loc["Σ", "i"] = ""
/tmp/ipykernel_181/2444747941.py:20: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  t1_display.loc["Σ", "x_i"] = ""
/tmp/ipykernel_181/2444747941.py:22: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '-' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  t1_display.loc["Σ", "u_i"] = "-"


,i,x_i,n_i,u_i,n_i*u_i,n_i*u_i^2,n_i*u_i^3,n_i*u_i^4,n_i*(u_i+1)^4
0,1.0,0.041563,2.0,-4.0,-8.0,32.0,-128.0,512.0,1.620000e+02
1,2.0,0.064687,2.0,-3.0,-6.0,18.0,-54.0,162.0,3.200000e+01
2,3.0,0.087813,16.0,-2.0,-32.0,64.0,-128.0,256.0,1.600000e+01
3,4.0,0.110937,12.0,-1.0,-12.0,12.0,-12.0,12.0,2.917038e-62
4,5.0,0.134063,29.0,0.0,0.0,0.0,0.0,0.0,2.900000e+01
5,6.0,0.157188,19.0,1.0,19.0,19.0,19.0,19.0,3.040000e+02
6,7.0,0.180312,21.0,2.0,42.0,84.0,168.0,336.0,1.701000e+03
7,8.0,0.203437,7.0,3.0,21.0,63.0,189.0,567.0,1.792000e+03
Σ,,,108.0,-,24.0,292.0,54.0,1864.0,4.036000e+03


### Контроль вычислений

Для восьмого столбца сумма, должна быть равна

$$
∑n_j (u_j+1)^4=∑n_j u_j^4 + 4∑n_j u_j^3 + 6 ∑n_j u_j^2 + 4∑n_j u_j + N
$$

In [ ]:
control_sum = sum_row["n_i*u_i^4"] + 4 * sum_row["n_i*u_i^3"] + 6 * sum_row["n_i*u_i^2"] + 4 * sum_row["n_i*u_i"] + N

control_sum

np.float64(4035.9999999999977)

Полученные значение совпадают

## Шаг 2. Условные и центральные эмпирические моменты
Условные моменты:
$$\bar M_k^* = \frac{1}{N}\sum n_i u_i^k.$$

Центральные моменты через $\bar M_k^*$ (для $k=2..4$):
$$\bar m_2 = (\bar M_2^*-(\bar M_1^*)^2)h^2,$$
$$\bar m_3 = (\bar M_3^* - 3\bar M_2^*\bar M_1^* + 2(\bar M_1^*)^3)h^3,$$
$$\bar m_4 = (\bar M_4^* - 4\bar M_3^*\bar M_1^* + 6\bar M_2^*(\bar M_1^*)^2 - 3(\bar M_1^*)^4)h^4.$$

In [ ]:

S1 = sum_row["n_i*u_i"]
S2 = sum_row["n_i*u_i^2"]
S3 = sum_row["n_i*u_i^3"]
S4 = sum_row["n_i*u_i^4"]

M1 = S1 / N
M2 = S2 / N
M3 = S3 / N
M4 = S4 / N

m2 = (M2 - M1**2)*h**2
m3 = (M3 - 3*M2*M1 + 2*(M1**3))*h**3
m4 = (M4 - 4*M3*M1 + 6*M2*(M1**2) - 3*(M1**4))*h**4

t2 = pd.DataFrame({
    "k": [1,2,3,4],
    "M*_k": [M1,M2,M3,M4],
    "m_k": [0, m2, m3, m4],
})
t2


,k,M*_k,m_k
0,1,0.222222,0.000000
1,2,2.703704,0.001419
2,3,0.500000,-0.000016
3,4,17.259259,0.000005


## Шаг 3–5. Среднее, дисперсия, СКО, асимметрия, эксцесс
$$\bar x = \frac{1}{N}∑x_i n_i = A + h\bar M_1^*,$$
$$D = m_2 = \frac{1}{N}∑n_i(x_i - \bar x)^2, \quad \sigma = \sqrt{\bar m_2}.$$

Исправленные оценки:
$$S^2 = \frac{N}{N-1}D, \quad S=\sqrt{S^2}.$$

In [ ]:
x_bar = sum(x_i*n_i)/N
x_bar_prime = C + h * M1

D = sum((x_i - x_bar) ** 2 * n_i ) / N
D_prime = m2
sigma = math.sqrt(m2)

S2_corr = (N/(N-1)) * D_prime if N > 1 else float("nan")
S_corr = math.sqrt(S2_corr) if N > 1 else float("nan")

As = m3 / (S_corr**3) if S_corr > 0 else float("nan")
Ex = (m4 / (S_corr**4) - 3) if S_corr > 0 else float("nan")

pd.DataFrame({
    "Показатель": ["N", "C                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ", "h", "x̄", "x̄'", "D", "D'",  "σ", "S² (испр.)", "S (испр.)", "A_s", "E_x"],
    "Значение":   [N, C, h, x_bar, x_bar_prime, D, D_prime, sigma, S2_corr, S_corr, As, Ex]
})


,Показатель,Значение
0,N,108.000000
1,C,0.134063
2,h,0.023125
3,x̄,0.139201
4,x̄',0.139201
5,D,0.001419
6,D',0.001419
7,σ,0.037675
8,S² (испр.),0.001433
9,S (испр.),0.037851


Выборочные среднее $x̄_в$ и дисперсия $D_в$, посчитанные по стандартной формуле и методом условных вариант, совпали, так как получились одинаковые значения $x̄ = 0.139201$ и $D = 0.001419$.

Исправленная дисперсия $s² = 0.001433$ и исправленное СКО $s = 0.037851$ немного больше смещённых оценок $D = 0.001419$ и $σ = 0.037675$, потому что при исправлении деление идёт на $N−1$, а не на $N$. Разница небольшая (примерно $+0.93\%$ по дисперсии и $+0.47\%$ по СКО), поэтому при $N = 108$ влияние смещения уже слабое.

Коэффициент асимметрии $A_s = −0.292$ показывает слабую левостороннюю асимметрию распределения. Эксцесс $E = −0.501$ отрицательный, значит распределение более плоское, чем нормальное (вершина ниже и хвосты легче).


## Шаг 6. Мода, медиана, коэффициент вариации

In [ ]:
# Вычисление моды
m_tilde = n_i / N

i = int(np.argmax(m_tilde))

# Нижняя граница модального интервала x_{i-1}
x_im1 = float(a[i])

mt_i   = float(m_tilde[i])
mt_im1 = float(m_tilde[i - 1]) if i > 0 else 0.0
mt_ip1 = float(m_tilde[i + 1]) if i < len(m_tilde) - 1 else 0.0

Mo = x_im1 + h * ((mt_i - mt_im1) / ((mt_i - mt_im1) + (mt_i - mt_ip1)))

# Вычисление медианы

m_tilde_cum = np.cumsum(m_tilde) # Накопленные относительные частоты
i = int(np.searchsorted(m_tilde_cum, 0.5, side="left"))
x_im1 = float(a[i])

# Сумма до предыдущего интервала
mt_i = float(m_tilde[i])
sum_prev = float(m_tilde_cum[i - 1]) if i > 0 else 0.0

Me = x_im1 + (h / mt_i) * (0.5 - sum_prev)

# Вычисление коэффициента вариации ---
V = (sigma / x_bar) * 100

pd.DataFrame({
    "Показатель": ["M_o", "M_e", "V"],
    "Значение": [Mo, Me, V]
})


,Показатель,Значение
0,M_o,0.137060
1,M_e,0.140043
2,V,27.065428


In [ ]:
# Границы интервала, где находится мода
a[i], a[i+1]

(np.float64(0.1225), np.float64(0.145625))

Мода меньше медианы, что указывает на правостороннюю асимметрию распределения: наиболее частое значение находится немного левее середины, а более крупные значения формируют «хвост» справа. Что отличается от оценки полученной по коэффициенту асимметрии. Это происходит, потому, как для оценки моды используется приблеженная формула, а расхожение возникает из-за группировки и округления, и выходит, что оценки моды (0.137), медианы (0.140) и среднего (0.139) близки к друг другу.

Коэффициент вариации отражает относительный уровень разброса. Полученное значение (V = 27%) свидетельствует о умеренной вариативности признака и однородности совокупности.

## Выводы

В ходе работы по заданным выборочным данным были получены точечные оценки основных характеристик распределения случайной величины методом моментов и методом условных вариант. Оценки среднего и дисперсии, найденные двумя способами, совпали, что подтверждает корректность вычислений и контрольных проверок. Исправленные оценки дисперсии и среднеквадратичного отклонения оказались немного больше смещённых из-за деления на N−1, при объёме выборки N=108 разница невелика и влияние смещения слабое. По коэффициенту асимметрии выявлена слабая левосторонняя асимметрия, а отрицательный эксцесс показывает более плоское распределение по сравнению с нормальным. Коэффициент вариации около 27% указывает на умеренный разброс и в целом достаточно однородную совокупность.
